packages/frontend/src/components/EpochTime.jsx

```diff
@@ -0,0 +1,37 @@
+import React, { useState, useEffect } from 'react'
+import { observer } from 'mobx-react-lite'
+import InfoCard from './InfoCard'
+import dayjs from 'dayjs'
+
+export default observer(({ deployment }) => {
+  const { startTimestamp, epochLength } = deployment
+  const timeSinceDeployment = new Date() / 1000 - startTimestamp
+  const [currentEpoch, setCurrentEpoch] = useState(0)
+  const [nextEpoch, setNextEpoch] = useState(0)
+  const calculateCurrentEpoch = () => {
+    const now = Math.floor(new Date() / 1000)
+    const current = Math.floor((now - startTimestamp) / epochLength)
+    const next = startTimestamp + epochLength * (current + 1)
+    setCurrentEpoch(current)
+    setNextEpoch(next)
+  }
+  useEffect(() => {
```

Comment:
This is another 1 off component that i'd rather see consolidated into the component that uses it.

The `calculateCurrentEpoch` logic doesn't really belong in a component. It _can_ work, but I think this component got created because the `calculateCurrentEpoch` logic was needed somewhere else. This is a really good function to put in maybe the unirep context. Make it like `calculateCurrentEpoch(deploymentId)` then components can use it very easily.

You can also move the callback logic below into the context. Something like `onEpochChange(deploymentId, callback)` would work. Then you could manage a single timer for each deployment.

Also, very important here, this effect hook needs a cleanup function. When the component is unmounted (like when the user navigates away) these timers need to be cancelled, otherwise the state info will continue to exist in memory because the timers depend on it. [Here](https://reactjs.org/docs/hooks-effect.html#effects-with-cleanup) is the documentation for `useEffects` cleanup (check the "Example Using Hooks").

So to cleanup here you'll need to keep track of the timers. This is another reason to prefer implementing in the context, you can track variables more easily/without using `useState`. If you do move the logic to the context make sure you have an `removeEpochChangeListener(id)` function that can be used to remove the subscription.

Reply:
similar to `EpochKeyInfo`, this one was created as a solution to a loading issue in `AttesterPage` - I didn't see a way to combine the useEffect logic when I moved `calculateCurrentEpoch` from the `EpochView` component into `AttesterPage`.
I will look into moving it into the unirep context and address the cleanup function.
How do you feel about merging this as is and working on these two issues in a new PR? Beacuse I'd like to get the other changes in here (footer, css, metadata fixes) updated asap since the app is live now.

Reply:
works for me